File: Day19/Part5/median_two_sorted_arrays.py
# LC 4 — Median of Two Sorted Arrays
# Binary search partition on the smaller array
from typing import List

def findMedianSortedArrays(nums1: List[int], nums2: List[int]) -> float:
    A, B = nums1, nums2
    if len(A) > len(B):
        A, B = B, A
    m, n = len(A), len(B)
    total = m + n
    half = total // 2
    lo, hi = 0, m
    while True:
        i = (lo + hi) // 2  # A partition size
        j = half - i
        Aleft = A[i-1] if i > 0 else float('-inf')
        Aright = A[i] if i < m else float('inf')
        Bleft = B[j-1] if j > 0 else float('-inf')
        Bright = B[j] if j < n else float('inf')
        if Aleft <= Bright and Bleft <= Aright:
            if total % 2:
                return float(min(Aright, Bright))
            return (max(Aleft, Bleft) + min(Aright, Bright)) / 2.0
        elif Aleft > Bright:
            hi = i - 1
        else:
            lo = i + 1

if __name__ == "__main__":
    print(findMedianSortedArrays([1,3],[2]))  # 2.0

File: Day19/Part5/median_two_sorted_arrays.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 4 — Median of Two Sorted Arrays
// Binary search partition on smaller array

double findMedianSortedArrays(vector<int>& A, vector<int>& B){
    if (A.size() > B.size()) return findMedianSortedArrays(B, A);
    int m = (int)A.size(), n = (int)B.size();
    int total = m + n;
    int half = total / 2;
    int lo = 0, hi = m;
    while(true){
        int i = (lo + hi) / 2;
        int j = half - i;
        int Aleft = (i > 0 ? A[i-1] : INT_MIN);
        int Aright = (i < m ? A[i] : INT_MAX);
        int Bleft = (j > 0 ? B[j-1] : INT_MIN);
        int Bright = (j < n ? B[j] : INT_MAX);
        if (Aleft <= Bright && Bleft <= Aright){
            if (total % 2) return min(Aright, Bright);
            return (max(Aleft, Bleft) + min(Aright, Bright)) / 2.0;
        } else if (Aleft > Bright){
            hi = i - 1;
        } else {
            lo = i + 1;
        }
    }
}

int main(){
    vector<int> a = {1,3}, b = {2};
    cout << fixed << setprecision(1) << findMedianSortedArrays(a,b) << "\n"; // 2.0
    return 0;
}
